File: BE/Sem 6/CNS/cipat/src/ts/standard_rsa/rsa.ts
/**
 * Standard RSA Implementation — No Optimizations
 *
 * Implements RSA from scratch using BigInt:
 *   - Key generation (random prime selection via Miller-Rabin)
 *   - Textbook RSA encrypt / decrypt (m^e mod n, c^d mod n)
 *   - Naive square-and-multiply modular exponentiation
 *   - PKCS#1 v1.5–style padding (simplified)
 *   - Hybrid mode: RSA for key-wrap + AES-256-GCM for bulk data
 *
 * No CRT, no pre-computation, no key caching, no batch shortcuts.
 */

import { randomBytes, createCipheriv, createDecipheriv } from "node:crypto";

// ---------------------------------------------------------------------------
// Modular exponentiation — naive square-and-multiply
// ---------------------------------------------------------------------------

/**
 * Square-and-multiply (left-to-right binary method).
 *
 * The *unoptimised* version: processes one bit at a time,
 * no windowing, no Montgomery reduction.
 */
export function modExp(base: bigint, exp: bigint, mod: bigint): bigint {
  let result = 1n;
  base = base % mod;
  while (exp > 0n) {
    if (exp & 1n) {
      result = (result * base) % mod;
    }
    exp >>= 1n;
    base = (base * base) % mod;
  }
  return result;
}

// ---------------------------------------------------------------------------
// Primality testing
// ---------------------------------------------------------------------------

/**
 * Miller-Rabin primality test with k rounds.
 */
export function isProbablePrime(n: bigint, k: number = 20): boolean {
  if (n < 2n) return false;
  if (n === 2n || n === 3n) return true;
  if (n % 2n === 0n) return false;

  // Write n-1 as 2^r * d
  let r = 0;
  let d = n - 1n;
  while (d % 2n === 0n) {
    d /= 2n;
    r++;
  }

  for (let round = 0; round < k; round++) {
    const a = randomBigIntInRange(2n, n - 2n);
    let x = modExp(a, d, n);
    if (x === 1n || x === n - 1n) continue;
    let found = false;
    for (let i = 0; i < r - 1; i++) {
      x = (x * x) % n;
      if (x === n - 1n) {
        found = true;
        break;
      }
    }
    if (!found) return false;
  }
  return true;
}

/**
 * Generate a random BigInt with the specified number of bits.
 */
function randomBigIntBits(bits: number): bigint {
  const bytes = Math.ceil(bits / 8);
  const buf = randomBytes(bytes);
  // Trim to exact bit length
  let result = bufToBigInt(buf);
  const mask = (1n << BigInt(bits)) - 1n;
  return result & mask;
}

/**
 * Generate a random BigInt in [min, max].
 */
function randomBigIntInRange(min: bigint, max: bigint): bigint {
  const range = max - min + 1n;
  const bits = range.toString(2).length;
  let result: bigint;
  do {
    result = randomBigIntBits(bits);
  } while (result > range);
  return result + min;
}

/**
 * Convert a Buffer to BigInt (big-endian).
 */
function bufToBigInt(buf: Buffer): bigint {
  let result = 0n;
  for (const b of buf) {
    result = (result << 8n) | BigInt(b);
  }
  return result;
}

/**
 * Generate a random prime of the specified bit length.
 */
export function generatePrime(bits: number): bigint {
  while (true) {
    let candidate = randomBigIntBits(bits);
    // Ensure high bit and odd
    candidate |= (1n << BigInt(bits - 1)) | 1n;
    if (isProbablePrime(candidate)) return candidate;
  }
}

// ---------------------------------------------------------------------------
// Key generation
// ---------------------------------------------------------------------------

function gcd(a: bigint, b: bigint): bigint {
  while (b) {
    [a, b] = [b, a % b];
  }
  return a;
}

/**
 * Modular inverse using extended Euclidean algorithm.
 */
function modInverse(e: bigint, phi: bigint): bigint {
  let [old_r, r] = [e, phi];
  let [old_s, s] = [1n, 0n];
  while (r !== 0n) {
    const q = old_r / r;
    [old_r, r] = [r, old_r - q * r];
    [old_s, s] = [s, old_s - q * s];
  }
  return ((old_s % phi) + phi) % phi;
}

export interface PublicKey {
  e: bigint;
  n: bigint;
}

export interface PrivateKey {
  d: bigint;
  n: bigint;
  p: bigint;
  q: bigint;
}

export type KeyPair = { priv: PrivateKey; pub: PublicKey };

/**
 * Generate RSA keypair. Each call generates fresh primes — no caching.
 */
export function generateKeypair(keySize: number = 2048): KeyPair {
  const half = keySize >> 1;
  let p = generatePrime(half);
  let q = generatePrime(half);
  while (q === p) {
    q = generatePrime(half);
  }

  const n = p * q;
  const phi = (p - 1n) * (q - 1n);
  let e = 65537n;
  while (gcd(e, phi) !== 1n) {
    e += 2n;
  }
  const d = modInverse(e, phi);

  return {
    priv: { d, n, p, q },
    pub: { e, n },
  };
}

// ---------------------------------------------------------------------------
// Encrypt / Decrypt
// ---------------------------------------------------------------------------

/** Textbook RSA encryption: c = m^e mod n. */
export function encrypt(pub: PublicKey, plaintext: bigint): bigint {
  return modExp(plaintext, pub.e, pub.n);
}

/** Textbook RSA decryption: m = c^d mod n (full-size exponent, no CRT). */
export function decrypt(priv: PrivateKey, ciphertext: bigint): bigint {
  return modExp(ciphertext, priv.d, priv.n);
}

// ---------------------------------------------------------------------------
// Byte-level helpers (padding)
// ---------------------------------------------------------------------------

/** Simplified PKCS#1 v1.5–style padding → BigInt. */
export function pkcs1Pad(data: Buffer, nBytes: number): bigint {
  const padLen = nBytes - data.length - 3;
  if (padLen < 8) throw new Error("Data too long for key size");
  const padding = randomBytes(padLen);
  // Ensure no zero bytes in padding
  for (let i = 0; i < padding.length; i++) {
    if (padding[i] === 0) padding[i] = 1;
  }
  const padded = Buffer.concat([
    Buffer.from([0x00, 0x02]),
    padding,
    Buffer.from([0x00]),
    data,
  ]);
  return bufToBigInt(padded);
}

/** Strip PKCS#1 v1.5 padding. */
export function pkcs1Unpad(paddedInt: bigint): Buffer {
  const hex = paddedInt.toString(16);
  const byteLen = Math.ceil(hex.length / 2);
  const data = Buffer.alloc(byteLen);
  // Write hex into buffer (big-endian)
  const paddedHex = hex.padStart(byteLen * 2, "0");
  for (let i = 0; i < byteLen; i++) {
    data[i] = parseInt(paddedHex.substring(i * 2, i * 2 + 2), 16);
  }
  // Skip 0x00 0x02 padding... 0x00
  let idx = 2;
  while (idx < data.length && data[idx] !== 0x00) idx++;
  idx++; // skip the 0x00 separator
  return data.subarray(idx);
}

export function encryptBytes(pub: PublicKey, data: Buffer): bigint {
  const nBytes = Math.ceil(pub.n.toString(2).length / 8);
  const m = pkcs1Pad(data, nBytes);
  return encrypt(pub, m);
}

export function decryptBytes(priv: PrivateKey, ciphertext: bigint): Buffer {
  const m = decrypt(priv, ciphertext);
  return pkcs1Unpad(m);
}

// ---------------------------------------------------------------------------
// Hybrid RSA + AES-256-GCM (for bulk data tests)
// Uses Node crypto for AES — RSA itself is from scratch.
// ---------------------------------------------------------------------------

export interface HybridEnvelope {
  wrappedKey: bigint;
  iv: Buffer;
  ct: Buffer;
  tag: Buffer;
}

export function hybridEncrypt(pub: PublicKey, data: Buffer): HybridEnvelope {
  const aesKey = randomBytes(32);
  const iv = randomBytes(12);

  const cipher = createCipheriv("aes-256-gcm", aesKey, iv);
  const ct = Buffer.concat([cipher.update(data), cipher.final()]);
  const tag = cipher.getAuthTag();

  const wrappedKey = encryptBytes(pub, aesKey);

  return { wrappedKey, iv, ct, tag };
}

export function hybridDecrypt(
  priv: PrivateKey,
  envelope: HybridEnvelope,
): Buffer {
  const aesKey = decryptBytes(priv, envelope.wrappedKey);

  const decipher = createDecipheriv(
    "aes-256-gcm",
    aesKey,
    envelope.iv,
  );
  decipher.setAuthTag(envelope.tag);
  return Buffer.concat([
    decipher.update(envelope.ct),
    decipher.final(),
  ]);
}

File: BE/Sem 6/CNS/cipat/src/standard_rsa/rsa.py
"""
Standard RSA Implementation — No Optimizations

Implements RSA from scratch using Python's built-in integers:
  - Key generation (random prime selection via Miller-Rabin)
  - Textbook RSA encrypt / decrypt (m^e mod n, c^d mod n)
  - Naive square-and-multiply modular exponentiation
  - PKCS#1 v1.5–style padding (simplified)
  - Hybrid mode: RSA for key-wrap + AES-256-CBC for bulk data

No CRT, no pre-computation, no key caching, no batch shortcuts.
"""

import random
import math
import os
import time
from typing import Tuple

# ---------------------------------------------------------------------------
# Primality testing
# ---------------------------------------------------------------------------

def is_probable_prime(n: int, k: int = 20) -> bool:
    """Miller-Rabin primality test with *k* rounds."""
    if n < 2:
        return False
    if n in (2, 3):
        return True
    if n % 2 == 0:
        return False

    # write n-1 as 2^r * d
    r, d = 0, n - 1
    while d % 2 == 0:
        d //= 2
        r += 1

    for _ in range(k):
        a = random.randrange(2, n - 1)
        x = pow(a, d, n)
        if x == 1 or x == n - 1:
            continue
        for _ in range(r - 1):
            x = pow(x, 2, n)
            if x == n - 1:
                break
        else:
            return False
    return True


def generate_prime(bits: int) -> int:
    """Generate a random prime of the specified bit length."""
    while True:
        candidate = random.getrandbits(bits)
        candidate |= (1 << (bits - 1)) | 1  # ensure high bit and odd
        if is_probable_prime(candidate):
            return candidate


# ---------------------------------------------------------------------------
# Modular exponentiation — naive square-and-multiply
# ---------------------------------------------------------------------------

def mod_exp(base: int, exp: int, mod: int) -> int:
    """Square-and-multiply (left-to-right binary method).

    This is the *unoptimised* version: processes one bit at a time,
    no windowing, no Montgomery reduction.
    """
    result = 1
    base = base % mod
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        exp >>= 1
        base = (base * base) % mod
    return result


# ---------------------------------------------------------------------------
# Key generation
# ---------------------------------------------------------------------------

def gcd(a: int, b: int) -> int:
    while b:
        a, b = b, a % b
    return a


def generate_keypair(key_size: int = 2048) -> Tuple[dict, dict]:
    """Return (private_key, public_key) dicts.

    Each call generates fresh primes — no caching.
    """
    half = key_size // 2
    p = generate_prime(half)
    q = generate_prime(half)
    while q == p:
        q = generate_prime(half)

    n = p * q
    phi = (p - 1) * (q - 1)

    e = 65537
    while gcd(e, phi) != 1:
        e += 2

    d = pow(e, -1, phi)  # modular inverse

    pub = {"e": e, "n": n}
    priv = {"d": d, "n": n, "p": p, "q": q}
    return priv, pub


# ---------------------------------------------------------------------------
# Encrypt / Decrypt
# ---------------------------------------------------------------------------

def encrypt(pub: dict, plaintext: int) -> int:
    """Textbook RSA encryption: c = m^e mod n."""
    return mod_exp(plaintext, pub["e"], pub["n"])


def decrypt(priv: dict, ciphertext: int) -> int:
    """Textbook RSA decryption: m = c^d mod n (full-size exponent, no CRT)."""
    return mod_exp(ciphertext, priv["d"], priv["n"])


# ---------------------------------------------------------------------------
# Byte-level helpers (padding)
# ---------------------------------------------------------------------------

def pkcs1_pad(data: bytes, n_bytes: int) -> int:
    """Simplified PKCS#1 v1.5–style padding → integer."""
    pad_len = n_bytes - len(data) - 3
    if pad_len < 8:
        raise ValueError("Data too long for key size")
    padding = bytes(random.randint(1, 255) for _ in range(pad_len))
    padded = b"\x00\x02" + padding + b"\x00" + data
    return int.from_bytes(padded, "big")


def pkcs1_unpad(padded_int: int) -> bytes:
    """Strip PKCS#1 v1.5 padding."""
    data = padded_int.to_bytes((padded_int.bit_length() + 7) // 8, "big")
    # skip 0x00 0x02 padding... 0x00
    idx = data.index(b"\x00", 2)
    return data[idx + 1:]


def encrypt_bytes(pub: dict, data: bytes) -> int:
    n_bytes = (pub["n"].bit_length() + 7) // 8
    m = pkcs1_pad(data, n_bytes)
    return encrypt(pub, m)


def decrypt_bytes(priv: dict, ciphertext: int) -> bytes:
    m = decrypt(priv, ciphertext)
    return pkcs1_unpad(m)


# ---------------------------------------------------------------------------
# Hybrid RSA + AES-256-GCM (for bulk data tests)
# Uses the cryptography library for AES — RSA itself is from scratch.
# This matches real-world practice where hybrid = custom RSA + library AES.
# ---------------------------------------------------------------------------

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes


def hybrid_encrypt(pub: dict, data: bytes) -> dict:
    """Hybrid: generate AES-256 key, encrypt data with AES-GCM, wrap AES key with RSA."""
    aes_key = os.urandom(32)
    iv = os.urandom(12)
    encryptor = Cipher(algorithms.AES(aes_key), modes.GCM(iv)).encryptor()
    ct = encryptor.update(data) + encryptor.finalize()
    wrapped_key = encrypt_bytes(pub, aes_key)
    return {"wrapped_key": wrapped_key, "iv": iv, "ct": ct, "tag": encryptor.tag}


def hybrid_decrypt(priv: dict, envelope: dict) -> bytes:
    """Hybrid: unwrap AES key via RSA, decrypt data with AES-GCM."""
    aes_key = decrypt_bytes(priv, envelope["wrapped_key"])
    decryptor = Cipher(algorithms.AES(aes_key), modes.GCM(envelope["iv"], envelope["tag"])).decryptor()
    return decryptor.update(envelope["ct"]) + decryptor.finalize()
